mersenne-31/src/poseidon2.rs

```diff
@@ -35,18 +35,21 @@ pub const MATRIX_DIAG_16_MERSENNE31: [Mersenne31; 16] =
 
 // We make use of the fact that most entries are a power of 2.
 // Note that this array is 1 element shorter than MATRIX_DIAG_16_MERSENNE31 as we do not include the first element.
-const MATRIX_DIAG_16_MONTY_SHIFTS: [i32; 15] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16];
+const MATRIX_DIAG_16_MONTY_SHIFTS: [u64; 15] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16];
 
 #[derive(Debug, Clone, Default)]
 pub struct DiffusionMatrixMersenne31;
 
 impl Permutation<[Mersenne31; 16]> for DiffusionMatrixMersenne31 {
+    #[inline]
     fn permute_mut(&self, state: &mut [Mersenne31; 16]) {
-        let sum: Mersenne31 = state.iter().cloned().sum();
-        state[0] = sum - state[0].double();
+        let part_sum: u64 = state.iter().skip(1).map(|x| x.value as u64).sum();
+        let full_sum = part_sum + (state[0].value as u64);
+        let s0 = part_sum + (Mersenne31::ORDER_U32 - state[0].value) as u64;
```

Comment:
I think the equivalent `(-state[i]).value` may be safest, in case we switch to an M31 impl where `value` can be 32 bits on certain targets